include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,401 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 - 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#  include <deal.II/base/conditional_ostream.h>
+#  include <deal.II/base/exceptions.h>
+#  include <deal.II/base/logstream.h>
+#  include <deal.II/base/mpi_stub.h>
+#  include <deal.II/base/parameter_handler.h>
+
+#  include <deal.II/lac/vector.h>
+#  include <deal.II/lac/vector_memory.h>
+
+#  include <boost/signals2.hpp>
+
+#  include <nvector/nvector_serial.h>
+
+#  include <deal.II/sundials/kinsol.h>
+
+#  include <deal.II/trilinos/nox.h>
+
+#  include <memory>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+template <typename VectorType = Vector<double>>
+class NLSolverSelector
+{
+public:
+
+  class AdditionalData
+    {
+    public:
+      /**
+       * KINSOL solution strategy. KINSOL includes a Newton-Krylov solver (both
+       * local and global) as well as Picard and fixed point solvers.
+       */
+      enum SolutionStrategy
+      {
+        /**
+         * Standard Newton iteration.
+         */
+        newton = KIN_NONE,
+        /**
+         * Newton iteration with linesearch.
+         */
+        linesearch = KIN_LINESEARCH,
+        /**
+         * Fixed point iteration.
+         */
+        fixed_point = KIN_FP,
+        /**
+         * Picard iteration.
+         */
+        picard = KIN_PICARD,
+      };
+
+      AdditionalData(const std::string &solver_type = "kinsol",
+                     const SolutionStrategy &strategy = linesearch,
+                     const unsigned int maximum_non_linear_iterations   = 200,
+                     const double       function_tolerance              = 0.0,
+                     const double       step_tolerance                  = 0.0,
+                     const bool         no_init_setup                   = false,
```

Comment:
I would not expose all options, like ``no_init_setup`` but use sensible defaults.